### PR TITLE
kOps: Fix and update upgrade tests

### PR DESCRIPTION
**Pull request by hakman**
/cc @rifelpet @johngmyers @olemarkus @justinsb

### CODE REVIEW

File: config/jobs/kubernetes/kops/build_jobs.py [hunk 1]
```diff
@@ -759,26 +759,9 @@ def generate_misc():
                        "--instance-manager=karpenter",
                        "--master-size=c6g.xlarge",
                    ],
-                   feature_flags=['Karpenter'],
                    extra_dashboards=["kops-misc"],
                    skip_regex=r'\[Slow\]|\[Serial\]|\[Disruptive\]|\[Flaky\]|\[Feature:.+\]|nfs|NFS|Gluster|Services.*rejected.*endpoints|TCP.CLOSE_WAIT|external.IP.is.not.assigned.to.a.node|same.port.number.but.different.protocols|same.hostPort.but.different.hostIP.and.protocol|should.create.a.Pod.with.SCTP.HostPort|Services.should.create.endpoints.for.unready.pods|Services.should.be.able.to.connect.to.terminating.and.unready.endpoints.if.PublishNotReadyAddresses.is.true|should.verify.that.all.nodes.have.volume.limits|In-tree.Volumes|LoadBalancers.should.be.able.to.preserve.UDP.traffic|serve.endpoints.on.same.port.and.different.protocols'), # pylint: disable=line-too-long
 
-        build_test(name_override="kops-aws-ipv6-karpenter",
```
Comment (johngmyers) [inline config/jobs/kubernetes/kops/build_jobs.py:766]: Why remove this test?

File: config/jobs/kubernetes/kops/build_jobs.py [hunk 2]
```diff
@@ -944,42 +927,36 @@ def generate_network_plugins():
...
         # 1.26 release branch
-        ((kops24, 'v1.20.6'), ('1.26', 'v1.21.7')),
-        ((kops24, 'v1.24.0'), ('1.26', 'v1.25.0')),
         ((kops25, 'v1.20.6'), ('1.26', 'v1.21.7')),
         ((kops25, 'v1.25.0'), ('1.26', 'v1.26.0')),
         ((kops26, 'v1.21.14'), ('1.26', 'v1.22.13')),
         ((kops26, 'v1.26.0'), ('1.26', 'v1.26.0')),
         # 1.27 release branch
-        ((kops25, 'v1.20.15'), ('1.27', 'v1.21.14')),
+        ((kops25, 'v1.20.15'), ('1.27', 'v1.22.17')),
```
Comment (johngmyers) [inline config/jobs/kubernetes/kops/build_jobs.py:942]: Does k8s 1.22 support kubelet 1.20?

File: config/jobs/kubernetes/kops/build_jobs.py [hunk 3]
```diff
@@ -944,42 +927,36 @@ def generate_network_plugins():
...
+        ((kops25, 'v1.21.14'), ('1.27', 'v1.22.17')),
+        ((kops25, 'v1.22.17'), ('1.27', 'v1.23.17')),
+        ((kops25, 'v1.23.17'), ('1.27', 'v1.24.15')),
+        ((kops25, 'v1.24.15'), ('1.27', 'v1.25.11')),
         ((kops25, 'v1.25.11'), ('1.27', 'v1.26.6')),
+        ((kops25, 'v1.26.6'), ('1.27', 'v1.27.3')),
```
Comment (johngmyers) [inline config/jobs/kubernetes/kops/build_jobs.py:948]: kops 1.25 and k8s 1.26 is not a valid combination, so this is an invalid test.

File: config/jobs/kubernetes/kops/build_jobs.py [hunk 4]
```diff
@@ -944,42 +927,36 @@ def generate_network_plugins():
...
         ((kops25, 'v1.25.0'), ('1.26', 'v1.26.0')),
         ((kops26, 'v1.21.14'), ('1.26', 'v1.22.13')),
         ((kops26, 'v1.26.0'), ('1.26', 'v1.26.0')),
         # 1.27 release branch
-        ((kops25, 'v1.20.15'), ('1.27', 'v1.21.14')),
+        ((kops25, 'v1.20.15'), ('1.27', 'v1.22.17')),
+        ((kops25, 'v1.21.14'), ('1.27', 'v1.22.17')),
+        ((kops25, 'v1.22.17'), ('1.27', 'v1.23.17')),
+        ((kops25, 'v1.23.17'), ('1.27', 'v1.24.15')),
+        ((kops25, 'v1.24.15'), ('1.27', 'v1.25.11')),
```
Comment (johngmyers) [inline config/jobs/kubernetes/kops/build_jobs.py:946]: I don't think it's worth testing all of these k8s versions in the middle.

File: config/jobs/kubernetes/kops/build_jobs.py [hunk 5]
```diff
@@ -944,42 +927,36 @@ def generate_network_plugins():
...
-        # 1.24 upgrade to latest
-        ((kops24, 'v1.23.1'), ('latest', 'v1.24.0')),
-        ((kops24, 'v1.24.0'), ('latest', 'v1.25.0')),
-        # 1.25 upgrade to latest
-        ((kops25, 'v1.24.0'), ('latest', 'v1.25.0')),
-        ((kops25, 'v1.25.0'), ('latest', 'v1.26.0')),
+        ((kops26, 'v1.22.17'), ('1.27', 'v1.23.17')),
+        ((kops26, 'v1.23.17'), ('1.27', 'v1.24.15')),
+        ((kops26, 'v1.24.15'), ('1.27', 'v1.25.11')),
+        ((kops26, 'v1.25.11'), ('1.27', 'v1.26.6')),
```
Comment (johngmyers) [inline config/jobs/kubernetes/kops/build_jobs.py:953]: I don't think it's worth testing all these k8s versions in the middle.

File: config/jobs/kubernetes/kops/build_jobs.py [hunk 6]
```diff
@@ -944,42 +927,36 @@ def generate_network_plugins():
...
+        ((kops26, 'v1.26.6'), ('1.27', 'v1.27.3')),
         # 1.26 upgrade to latest
-        ((kops26, 'v1.21.14'), ('latest', 'v1.22.1')),
+        ((kops26, 'v1.21.0'), ('latest', 'v1.23.0')),
```
Comment (johngmyers) [inline config/jobs/kubernetes/kops/build_jobs.py:956]: Does k8s 1.23 support kubelet 1.22?

File: config/jobs/kubernetes/kops/build_jobs.py [hunk 7]
```diff
@@ -944,42 +927,36 @@ def generate_network_plugins():
...
+        ((kops26, 'v1.22.0'), ('latest', 'v1.23.0')),
+        ((kops26, 'v1.23.0'), ('latest', 'v1.24.0')),
         ((kops26, 'v1.24.0'), ('latest', 'v1.25.0')),
```
Comment (johngmyers) [inline config/jobs/kubernetes/kops/build_jobs.py:959]: I don't think it's worth testing these k8s versions in the middle.

**Review by johngmyers — CHANGES_REQUESTED** [review]

File: config/jobs/kubernetes/kops/build_jobs.py [hunk 8]
```diff
@@ -944,42 +927,36 @@ def generate_network_plugins():
...
-        ((kops26, 'v1.25.0'), ('latest', 'v1.25.0')),
         ((kops26, 'v1.25.0'), ('latest', 'v1.26.0')),
```
Comment (johngmyers) [inline config/jobs/kubernetes/kops/build_jobs.py:960]: Also add a k8s 1.26 to k8s 1.26 upgrade

File: config/jobs/kubernetes/kops/build_jobs.py [hunk 9]
```diff
@@ -1611,13 +1587,35 @@ def generate_presubmits_e2e():
...
+                'KOPS_VERSION_A': "1.26",
+                'K8S_VERSION_A': "v1.26.0",
+                'KOPS_VERSION_B': "latest",
+                'K8S_VERSION_B': "v1.27.0",
+                'KOPS_SKIP_E2E': '1',
+                'KOPS_CONTROL_PLANE_SIZE': '3',
+            }
+        ),
+        presubmit_test(
+            name="pull-kops-e2e-aws-upgrade-k126-ko127-to-k127-kolatest-karpenter",
```
Comment (johngmyers) [inline config/jobs/kubernetes/kops/build_jobs.py:1613]:
```suggestion
            name="pull-kops-e2e-aws-upgrade-k127-ko127-to-k127-kolatest-karpenter",
```
?

File: config/jobs/kubernetes/kops/build_jobs.py (hunk 1 above)
Comment (hakman) [reply config/jobs/kubernetes/kops/build_jobs.py:766]: This fails to even start.

File: config/jobs/kubernetes/kops/build_jobs.py (hunk 2 above)
Comment (hakman) [reply config/jobs/kubernetes/kops/build_jobs.py:942]: Not sure I get the Q

File: config/jobs/kubernetes/kops/build_jobs.py (hunk 3 above)
Comment (hakman) [reply config/jobs/kubernetes/kops/build_jobs.py:948]: True, fixed.

File: config/jobs/kubernetes/kops/build_jobs.py (hunk 4 above)
Comment (hakman) [reply config/jobs/kubernetes/kops/build_jobs.py:946]: Removed

File: config/jobs/kubernetes/kops/build_jobs.py (hunk 5 above)
Comment (hakman) [reply config/jobs/kubernetes/kops/build_jobs.py:953]: Removed

File: config/jobs/kubernetes/kops/build_jobs.py (hunk 2 above)
Comment (hakman) [reply config/jobs/kubernetes/kops/build_jobs.py:942]: It is an experiment.

File: config/jobs/kubernetes/kops/build_jobs.py (hunk 6 above)
Comment (hakman) [reply config/jobs/kubernetes/kops/build_jobs.py:956]: It is an experiment.

File: config/jobs/kubernetes/kops/build_jobs.py (hunk 7 above)
Comment (hakman) [reply config/jobs/kubernetes/kops/build_jobs.py:959]: Removed.

File: config/jobs/kubernetes/kops/build_jobs.py (hunk 8 above)
Comment (hakman) [reply config/jobs/kubernetes/kops/build_jobs.py:960]: Done

File: config/jobs/kubernetes/kops/build_jobs.py (hunk 9 above)
Comment (hakman) [reply config/jobs/kubernetes/kops/build_jobs.py:1613]: This is intentional, want to see some nodes rolling also.

File: config/jobs/kubernetes/kops/build_jobs.py (hunk 1 above)
Comment (johngmyers) [reply config/jobs/kubernetes/kops/build_jobs.py:766]: It's useful signal. We should create issues to fix Karpenter on IPv6 clusters.

File: config/jobs/kubernetes/kops/build_jobs.py (hunk 2 above)
Comment (johngmyers) [reply config/jobs/kubernetes/kops/build_jobs.py:942]: Ah, the [version skew policy](https://kubernetes.io/releases/version-skew-policy/) still states that kube-apiserver only supports a skew of one version. So an HA cluster can only upgrade one k8s version at a time. The supported version skew between kubelet and kube-apiserver is two, so a single-control-plane-node cluster could theoretically upgrade two k8s versions, though I don't think this is worth testing.

If it is an experiment, it should be commented as such so we don't keep propagating it forward.

Comment (hakman) [reply config/jobs/kubernetes/kops/build_jobs.py:942]: Removed

File: config/jobs/kubernetes/kops/build_jobs.py (hunk 1 above)
Comment (hakman) [reply config/jobs/kubernetes/kops/build_jobs.py:766]: Added back

File: config/jobs/kubernetes/kops/build_jobs.py [hunk 10]
```diff
@@ -944,49 +942,30 @@ def generate_network_plugins():
...
-        ((kops25, 'v1.25.0'), ('1.25', 'v1.25.0')),
         # 1.26 release branch
-        ((kops24, 'v1.20.6'), ('1.26', 'v1.21.7')),
-        ((kops24, 'v1.24.0'), ('1.26', 'v1.25.0')),
         ((kops25, 'v1.20.6'), ('1.26', 'v1.21.7')),
         ((kops25, 'v1.25.0'), ('1.26', 'v1.26.0')),
         ((kops26, 'v1.21.14'), ('1.26', 'v1.22.13')),
         ((kops26, 'v1.26.0'), ('1.26', 'v1.26.0')),
         # 1.27 release branch
-        ((kops25, 'v1.20.15'), ('1.27', 'v1.21.14')),
```
Comment (johngmyers) [inline config/jobs/kubernetes/kops/build_jobs.py:967]: This should be kops 1.25 k8s 1.21 to kops 1.27 k8s 1.22

File: config/jobs/kubernetes/kops/build_jobs.py [hunk 11]
```diff
@@ -944,49 +942,30 @@ def generate_network_plugins():
...
         ((kops25, 'v1.25.11'), ('1.27', 'v1.26.6')),
         ((kops26, 'v1.21.14'), ('1.27', 'v1.22.17')),
-        ((kops26, 'v1.26.6'), ('1.27', 'v1.26.6')),
-        ((kops27, 'v1.22.17'), ('1.27', 'v1.23.17')),
-        ((kops27, 'v1.27.3'), ('1.27', 'v1.27.3')),
```
Comment (johngmyers) [inline config/jobs/kubernetes/kops/build_jobs.py:972]: These two are still relevant and should be kept

File: config/jobs/kubernetes/kops/build_jobs.py [hunk 12]
```diff
@@ -944,49 +942,30 @@ def generate_network_plugins():
...
-        # 1.24 upgrade to latest
-        ((kops24, 'v1.23.1'), ('latest', 'v1.24.0')),
-        ((kops24, 'v1.24.0'), ('latest', 'v1.25.0')),
-        # 1.25 upgrade to latest
-        ((kops25, 'v1.24.0'), ('latest', 'v1.25.0')),
-        ((kops25, 'v1.25.0'), ('latest', 'v1.26.0')),
```
Comment (johngmyers) [inline config/jobs/kubernetes/kops/build_jobs.py:978]: These three are still relevant and should be kept.

**Review by johngmyers — CHANGES_REQUESTED** [review]

File: config/jobs/kubernetes/kops/build_jobs.py (hunk 10 above)
Comment (hakman) [reply config/jobs/kubernetes/kops/build_jobs.py:967]: Done

File: config/jobs/kubernetes/kops/build_jobs.py (hunk 11 above)
Comment (hakman) [reply config/jobs/kubernetes/kops/build_jobs.py]: Done

File: config/jobs/kubernetes/kops/build_jobs.py (hunk 12 above)
Comment (hakman) [reply config/jobs/kubernetes/kops/build_jobs.py]: Done

**Review by johngmyers — APPROVED** [review]

**Pull request merged**